In [5]:
import numpy as np
import pandas as pd

## Question 1

In [4]:
np.__version__

'1.21.5'

In [6]:
pd.__version__

'1.4.3'

### Loading the Data

In [7]:
df = pd.read_csv("data.csv")

In [11]:
df.head(10)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   
5  BMW    1 Series  2012  premium unleaded (required)      230.0   
6  BMW    1 Series  2012  premium unleaded (required)      300.0   
7  BMW    1 Series  2012  premium unleaded (required)      300.0   
8  BMW    1 Series  2012  premium unleaded (required)      230.0   
9  BMW    1 Series  2013  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   
5               6.0            MANUAL  rear wheel drive              2.0   
6               6.0            MANUAL  rear wheel drive              2.0   
7               6.0            MANUAL  rear wheel drive              2.0   
8               6.0            MANUAL  rear wheel drive              2.0   
9               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   
5                     Luxury,Performance      Compact         Coupe   
6                     Luxury,Performance      Compact   Convertible   
7                Luxury,High-Performance      Compact         Coupe   
8                                 Luxury      Compact   Convertible   
9                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500  
5           28        18        3916  31200  
6           26        17        3916  44100  
7           28        20        3916  39300  
8           28        18        3916  36900  
9           27        18        3916  37200

## Question 2
How many records are in the dataset?

In [12]:
df.shape

(11914, 16)

the number of records in the dataset are 11914

## Question 3
Who are the most frequent car manufacturers (top-3) according to the dataset?

In [18]:
df['Make'].value_counts()[:3]

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

## Question 4
What's the number of unique Audi car models in the dataset?

In [30]:
audi_car = df[df["Make"] == 'Audi']["Model"].nunique()
audi

34

## Question 5
How many columns in the dataset have missing values?

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

The number of columns in the dataset have missing values = 5 (Engine Fuel Type - Engine HP - Engine Cylinders - Number of Doors - Market Category)

## Question 6
Find the median value of "Engine Cylinders" column in the dataset.

In [35]:
df["Engine Cylinders"].median()

6.0

Next, calculate the most frequent value of the same "Engine Cylinders".

In [37]:
df["Engine Cylinders"].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

In [45]:
df["Engine Cylinders"].fillna(float(df["Engine Cylinders"].mode()), inplace =True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11914 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

Now, calculate the median value of "Engine Cylinders" once again.
Has it changed?

In [46]:
df["Engine Cylinders"].median()

6.0

No, it is not changed

## Question 7
Select all the "Lotus" cars from the dataset.

In [52]:
lotus = df[df["Make"] == "Lotus"]
lotus

Make      Model  Year                Engine Fuel Type  Engine HP  \
3912  Lotus      Elise  2009     premium unleaded (required)      189.0   
3913  Lotus      Elise  2009     premium unleaded (required)      218.0   
3914  Lotus      Elise  2009     premium unleaded (required)      189.0   
3915  Lotus      Elise  2010     premium unleaded (required)      189.0   
3916  Lotus      Elise  2010     premium unleaded (required)      218.0   
3917  Lotus      Elise  2011     premium unleaded (required)      189.0   
3918  Lotus      Elise  2011     premium unleaded (required)      217.0   
3919  Lotus      Elise  2011     premium unleaded (required)      217.0   
4216  Lotus     Esprit  2002     premium unleaded (required)      350.0   
4217  Lotus     Esprit  2003     premium unleaded (required)      350.0   
4218  Lotus     Esprit  2004     premium unleaded (required)      350.0   
4257  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4258  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4259  Lotus      Evora  2011     premium unleaded (required)      276.0   
4260  Lotus      Evora  2011     premium unleaded (required)      276.0   
4261  Lotus      Evora  2013     premium unleaded (required)      276.0   
4262  Lotus      Evora  2013     premium unleaded (required)      345.0   
4263  Lotus      Evora  2013     premium unleaded (required)      345.0   
4264  Lotus      Evora  2013     premium unleaded (required)      276.0   
4265  Lotus      Evora  2014     premium unleaded (required)      345.0   
4266  Lotus      Evora  2014     premium unleaded (required)      276.0   
4267  Lotus      Evora  2014     premium unleaded (required)      345.0   
4268  Lotus      Evora  2014     premium unleaded (required)      276.0   
4292  Lotus      Exige  2009  premium unleaded (recommended)      257.0   
4293  Lotus      Exige  2009  premium unleaded (recommended)      240.0   
4294  Lotus      Exige  2010  premium unleaded (recommended)      240.0   
4295  Lotus      Exige  2011  premium unleaded (recommended)      240.0   
4296  Lotus      Exige  2011  premium unleaded (recommended)      257.0   
4297  Lotus      Exige  2011  premium unleaded (recommended)      257.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   
3917               4.0            MANUAL  rear wheel drive              2.0   
3918               4.0            MANUAL  rear wheel drive              2.0   
3919               4.0            MANUAL  rear wheel drive              2.0   
4216               8.0            MANUAL  rear wheel drive              2.0   
4217               8.0            MANUAL  rear wheel drive              2.0   
4218               8.0            MANUAL  rear wheel drive              2.0   
4257               6.0            MANUAL  rear wheel drive              2.0   
4258               6.0            MANUAL  rear wheel drive              2.0   
4259               6.0            MANUAL  rear wheel drive              2.0   
4260               6.0            MANUAL  rear wheel drive              2.0   
4261               6.0            MANUAL  rear wheel drive              2.0   
4262               6.0            MANUAL  rear wheel drive              2.0   
4263               6.0            MANUAL  rear wheel drive              2.0   
4264               6.0            MANUAL  rear wheel drive              2.0   
4265               6.0            MANUAL  rear wheel drive              2.0   
4266               6.0            MANUAL  rear wheel drive              2.0   
4267              

Select only columns "Engine HP", "Engine Cylinders".


In [59]:
EHP_EC = lotus[["Engine HP", "Engine Cylinders"]]
EHP_EC

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0
3917      189.0               4.0
3918      217.0               4.0
3919      217.0               4.0
4216      350.0               8.0
4217      350.0               8.0
4218      350.0               8.0
4257      400.0               6.0
4258      400.0               6.0
4259      276.0               6.0
4260      276.0               6.0
4261      276.0               6.0
4262      345.0               6.0
4263      345.0               6.0
4264      276.0               6.0
4265      345.0               6.0
4266      276.0               6.0
4267      345.0               6.0
4268      276.0               6.0
4292      257.0               4.0
4293      240.0               4.0
4294      240.0               4.0
4295      240.0               4.0
4296      257.0               4.0
4297      257.0               4.0

Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).


In [66]:
EHP_EC = EHP_EC.drop_duplicates()
EHP_EC.shape

(9, 2)

Get the underlying NumPy array. Let's call it X.


In [68]:
X = EHP_EC.to_numpy()
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.


In [71]:
XT = X.T
XT

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [86]:
XTX = np.dot(XT, X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

Invert XTX.


In [87]:
IXTX = np.linalg.inv(XTX)
IXTX

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].


In [88]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])


Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.


In [90]:
w = np.dot(np.dot(IXTX,XT), y.T)
w

array([  4.59494481, -63.56432501])

What's the value of the first element of w?

The Value of the first element of w = 4.59494481